Parallel pathway file generation

johnbradley opened the pull request:
- Changes `generate_depmap_pathways.R` to build only a subset of the positive or negative data.
- Adds `merge_depmap_pathways.R` to merge the subset files into either a master_positive or master_negative RData file. 
- Adds `depmap_pathways_util.R` that abstracts out the common parts shared between `generate_depmap_pathways.R` and `merge_depmap_pathways.R`
- Changes the Makefile to accommodate running these R scripts. To execute in parallel requires running make with the [parallel execution flag: -j <number_workers>](https://www.gnu.org/software/make/manual/html_node/Parallel.html). 
- Changes build-slurm.sh [to generate the depmap_pathways data with 10 processes](https://github.com/Duke-GCB/ddh/blob/ea86b7c684f2a83ecae71b3489b45f46060dcf81/build-slurm.sh#L13).
- Includes change to `generate_depmap_pathways.R` enrichr_loop function to an error due to Issue #18.

We currently generate 20 subset files:
- positive_subset_1_of_10.Rds thru positive_subset_10_of_10.Rds to build master_positive.RData
- negative_subset_1_of_10.Rds thru negative_subset_10_of_10.Rds to build master_negative.RData

Fixes #16 
Fixes #18

## Note
This PR significantly changes `generate_depmap_pathways.R` into many small functions. `generate_depmap_pathways.R` is expected to be run from the command line. It only runs the `main()` function when not interactive(RStudio) to allow easier debugging. From within RStudio you can run the entire script and exercise specific functions later.

This PR also introduces a utility R script(depmap_pathways_util.R) shared between multiple R scripts. This is similar to what we do with `current_release.R` but includes functions. The functions inside depmap_pathways_util.R all start with `dpu_`. This doesn't seem typical for R so perhaps I should drop the `dpu_` prefix?

This PR should probably wait until #17 merges since there is some overlap.

hirscheylab commented:
> All the examples I have seen this is only a problem for the bottom pathway. The genes in question had some correlated genes for the top pathway. @hirscheylab I could email you the full list of genes that had this problem if that would help.

@johnbradley : Please send the list of zero genes when you can; no rush.

hirscheylab commented:
Conflicts with 3 minor changes I made to the system timer; not critical. Just thought it'd be good to know how long each of the steps take.

johnbradley commented on code/generate_depmap_pathways.R:
```diff
@@ -29,104 +36,130 @@ enrichr_loop <- function(gene_list, databases){
+      select(approved_symbol, approved_name, fav_gene) %>%
+      rename(gene = approved_symbol, name = approved_name, r2 = fav_gene) %>%
+      pull("gene") %>%
+      c(fav_gene, .) %>%
```
The above line is the only difference between Top pathway enrichment and Bottom pathway enrichment.
Is this intentional. This might be why the top pathway enrichment doesn't fail in enrichr_loop.

hirscheylab replied:
Perhaps...that would give a gene list with only 1 gene. The reason that is combine fav_gene with the gene list is because for the positively correlated genes, I want to append the query gene into the list to generate pathways that *include* the query gene. This would be equivalent to an r^2 == 1. But this is not the case for the negative correlated genes. 

So: yes, the inclusion of c(fav_gene, .) is intentional to be only in the top enrichment pathways

johnbradley commented on build-slurm.sh:
```diff
@@ -1,10 +1,13 @@
-RSCRIPT_CMD="singularity exec singularity/images/depmap.sif Rscript" make
+export RSCRIPT_CMD="singularity exec singularity/images/depmap.sif Rscript"
+make gene_summary depmap_data depmap_stats depmap_tables
+make -j 10 depmap_pathways
```
I had intended for these two lines to merely be
```
make -j 10
```
There are issues running the earlier rules with multiple jobs: A process is spawned for each target of the rule so for example we end up running 5 copies of `code/generate_depmap_data.R`. I will create an issue to address this.

johnbradley replied:
I created #25 to address this.

johnbradley commented:
> But I think one of the R-focused parallelization strategies (such as those mentioned in #16) or forEach/doParallel would be a better fit.

@dleehr If we don't create intermediate files, is the plan to keep retrying the entire 6 hour process until it succeeds? I have had to run `code/create_gene_summary.R` several times to get a successful run. I am concerned we may see similar sporadic issues in the long term with this code.

I briefly looked at `furrr`. It worked fine for simple examples but the background processes just hung when trying to  run our code in `code/generate_depmap_data.R`. I will look at forEach/doParallel.

johnbradley commented:
I moved the pathway data generation parallelism into a shell script and simplified the Makefile so we only have one place to change the number of intermediate files used for pathway data generation.
